### PR TITLE
lang: Protect against Bus overwrites where possible

### PR DESCRIPTION
**Pull request by scztt**
When a Bus object is provided as the argument for an Out, we have an opportunity to validate that we're outputting a reasonable number of channels - we should do that. Outputting more than the size of your bus transparently overflows into whatever bus is next numerically, which can cause hard to diagnose bugs.

### CODE REVIEW

**Comment by scztt** [comment]
This is a simple one, but this bit me for the nth time today. It's ALMOST better to do this for anything that `.respondsTo(\numChannels)`... except that Object:numChannels returns 1 by default, which means warnings in the very very common case where you use a hardcoded integer as the arg for an Out.

**Comment by jamshark70** [comment]
To do this properly is trickier than it sounds. The normal (I'd almost say "correct") first input to Out is a synth control -- `Out.ar(out, ...)` -- rather than a hardcoded index or Bus object. It's theoretically possible for SynthDesc to identify which control corresponds to the output (it already tries to do this, but it isn't reliable -- it's easily confused by a LocalOut, e.g.), and given that, it's theoretically possible for Synth.newMsg and for Event types to look up the SynthDesc, scan the argument list, check a corresponding Bus object against the output spec and warn. But, making sure that this is trustworthy *and* doesn't kill performance seems like it would be difficult.

I just had a question on the mailing list the other day about playing a synth on a MixerChannel, where it didn't work because the user had hardcoded a bus number... and I didn't go so far at the time, but I was tempted to say that the documentation should recommend against hardcoding (i.e., that a control input should generally be preferred). If there's consensus in favor of that recommendation, it means this check would do nothing in the typical case... in which case, I'd be somewhat skeptical of putting it in the main class library (because, if it's in main, then it should be documented, but if we advertise to users that bus sizes will be checked, then they should be checked in the common case, not only in an unusual corner case that is counter-recommended in the documentation).

So... I don't know about this one. I see the point, but I'm not sure how generally applicable it is.

Also, shouldn't the wording be: "% is writing % channels out to a bus of size %"?

**Comment by brianlheim** [comment]
> So... I don't know about this one. I see the point, but I'm not sure how generally applicable it is.

IMO, it's totally OK for a warning/info message to cover a common case without covering the general case.

I also hardcode my bus indices (using global environment variables) because my control flows are fairly simple.

Even if we gave a recommendation not to hardcode, I suspect that this warning message would still be helpful while debugging. I think the sanity check "if I set this directly to N, does it work?" is a frequent one, and this warning message would catch programming errors early at that point.

So, :+1: from me.

**Comment by jamshark70** [comment]
I don't object to the warning. I'm a little concerned that warning about this case, without warning about other cases, might give users the impression that it's *better* to hardcode Bus objects into `Out` units specifically to take advantage of the warning.

If that becomes a common view, then I would start getting questions like this:

> I created a MixerChannel and SynthDef like this:
> 
> ```
> m = MixerChannel(\test, s, 2, 2);
>
> SynthDef(\test, { ...
>     ...
>     Out.ar(m.inbus, ...);
> }).add;
> ```
> 
> Now I want to create a second mixer `n = MixerChannel(\test2, s, 2, 2)` and use the `\test` SynthDef with it... but if I do `n.play(\test)`, the sound goes through `m` instead.

Well, that hypothetical question is because of picking up a ~bad~ inflexible SynthDef habit somewhere.

It's for this reason that I'm really tempted to say hardcoding `Out.ar(aBus, ...)` is an anti-pattern. Even Brian's defense of it admits that getting away with it depends on simple signal flows -- but the moment you say e.g. "Pfxb would be perfect for this case," then it no longer applies and you have to go back and change your code (where it's not really any more difficult to write the code in the first place so that it can adapt to different situations).

TL;DR I'm fine with adding the warning, but I'm not sure it's *correct*.

**Comment by jamshark70** [comment]
Alternately, provide an interface to support other kinds of Bus-like objects. To wit:

```
s.boot;

m = MixerChannel(\test, s, 1, 2);
-> MixerChannel(test, localhost, 1, 2)

m.inChannels
-> 1

m.inbus.numChannels
-> 2
```

The mixer expects 1 in, and provides 2 out, but it needs to allocate the larger number of channels for its internal bus. So `Out.ar(m.inbus, [left, right])` would not function properly, but it would not trigger the warning.

Sorry to be persnickety but it's worth covering more bases.

**Comment by jamshark70** [comment]
Demonstrating the point 😉 -- http://new-supercollider-mailing-lists-forums-use-these.2681727.n2.nabble.com/multiple-channels-in-ndefs-and-synthdefs-td7632028.html

> > //8 channel output sinewave synthdef
> > (
> > SynthDef(\eightsine, {|freq|
> > var sig = PanAz.ar(8, SinOsc.ar(freq), LFDNoise1.ar(2));
> > Out.ar(0, sig);
> > }).add;
> > )
> 
> That's not ideal - the secret behind Ndefs, NodeProxies, ProxySpaces is
> that they take care of their out bus themselves. So you should never
> hardcode the out bus in a SynthDef that you're gonna use in an Ndef.

Sometime later (I have class now), I'll grep the HelpSource directory for `Out.ar(0...)`... IMO these should be changed to `Out.ar(out, ...)` because new users don't necessarily understand right away (not their fault, it's not easy to learn any kind of programming) -- they copy from examples, including bad habits in the examples.

That's a semi-OT remark, but while the subject is open...

File: SCClassLibrary/Common/Audio/InOut.sc [hunk 1]
```diff
@@ -237,10 +249,12 @@ AbstractOut : UGen {
 Out : AbstractOut {
 	*ar { arg bus, channelsArray;
 		channelsArray = this.replaceZeroesWithSilence(channelsArray.asUGenInput(this).asArray);
+		this.checkChannelCount(bus, channelsArray);
```
Comment (telephon) [inline SCClassLibrary/Common/Audio/InOut.sc]: this will never be called, because the line before:
`this.replaceZeroesWithSilence(channelsArray.asUGenInput(this).asArray);`
already has converted the bus to its index.

File: SCClassLibrary/Common/Audio/InOut.sc [hunk 2]
```diff
@@ -271,11 +285,13 @@ LocalOut : AbstractOut {
 XOut : AbstractOut {
 	*ar { arg bus, xfade, channelsArray;
 		channelsArray = this.replaceZeroesWithSilence(channelsArray.asUGenInput(this).asArray);
-		this.multiNewList(['audio', bus, xfade] ++ channelsArray)
+		this.checkChannelCount(bus, channelsArray);
```
Comment (telephon) [inline SCClassLibrary/Common/Audio/InOut.sc]: same problem here.

**Comment by brianlheim** [comment]
This hasn't received an update since last year, so I'm closing it for now.
